Applications/Spire/Include/Spire/Spire/QtFuture.hpp

```diff
@@ -5,34 +5,66 @@
 #include <Beam/Routines/Async.hpp>
 #include <boost/noncopyable.hpp>
 #include "Spire/Spire/QtPromise.hpp"
-#include "Spire/Spire/Spire.hpp"
 
 namespace Spire {
 
   /*! \brief Encapsulates a QtPromise resolving to a future value.
       \tparam T The type of value to resolve to.
   */
   template<typename T>
-  class QtFuture : private boost::noncopyable {
+  class BaseQtFuture : private boost::noncopyable {
     public:
 
       //! The type of value to resolve to.
       using Type = T;
 
-      QtFuture(QtFuture&&) = default;
+      //! Resolves the QtPromise waiting on this future to an exception.
+      void resolve(std::exception_ptr e);
+
+    protected:
+      Beam::Routines::Eval<Type> m_eval;
+
+      BaseQtFuture(Beam::Routines::Eval<Type> eval);
+  };
+
+  /*! \brief Encapsulates a QtPromise resolving to a future value.
+      \tparam T The type of value to resolve to.
+  */
+  template<typename T>
+  class QtFuture : public BaseQtFuture<T> {
+    public:
+
+      QtFuture(QtFuture&&);
 
-      QtFuture& operator =(QtFuture&&) = default;
+      QtFuture& operator =(QtFuture&&);
```

Comment:
Inheritance doesn't conflict with default move operators, what happens specifically if you default the move operators?

Reply:
I am not sure about the move assignment - makes sense if it can be defaulted - but, for a move constructor in a derived class, I have to explicitly call the base's move constructor. It kind of makes sense as the base class might have a constructor taking an instance of the derived class, making defaulting stuff like:
```
Derived(Derived&& other)
  : Base(std::move(other)),
    m_member(std::move(other).m_member) {}
```
unsafe.
I can default BaseQtFuture's move constructor and call it in QtFuture's move constructors, yet I thought it would be unnecessary as `m_eval` is transparent to QtFuture.

If I do default both BaseQtFuture's and QtFuture's move constructors, the latter will default to deleted, making moving instances of QtFuture impossible. As a result, I would not be able to call `std::vector<QtFuture>::push_back` either.

Reply:
>but, for a move constructor in a derived class, I have to explicitly call the base's move constructor.

I don't follow, why do you need to explicitly call the base move constructor at all? Default the move constructors so nothing is called explicitly.

Reply:
Well, the compilation failure only occurs when inheriting from `boost::noncopyable`. For example, the following code:
```
  struct A {
    A(int y) : x(y) {}
    A(A&&) = default;

    int x;
  };

  struct B : public A {
    B(int y) : A(y) {}
    B(B&&) = default;
  };

  void test() {
    auto b1 = B(5);
    auto b2 = std::move(b1);
  }
```
compiles fine, but as soon as A is a subclass of `boost::noncopyable` VS compiler will render B's defaulted move constructor deleted. Now I need to figure out whether it is an issue with the compiler, `boost::noncopyable`, or wrong use of `boost::noncopyable` on our side.

Reply:
Okay remove the use of ```boost::noncopyable``` then and delete the copy constructors explicitly. The less boost, the better. Also use private inheritance on BaseQtPromise and inherit its constructors in QtFuture instead of reimplementing them.

Reply:
Copy constructors are implicitly deleted as long as there are explicit move constructors. Also, `Eval` is noncopyable too, making any would-be defined copy operation ill-formed. Should I still explicitly delete the copies for coding style or anything?

Reply:
No that's fine then.